File: questions/objectDiff/objectdiff.js
const Mocha = require('mocha')
const { expect } = require('chai')
const mocha = new Mocha()
mocha.suite.emit('pre-require', this, 'solution', mocha)

const diff = (before, after) => {
  // write your implementation here!
  throw "Not implemented";
}


describe('Object differ test case', () => {

  it('should show diff for added and removed elements', () => {
    const before = {
      "hello_world.js": [516, "javascript", "Oct 12 09:04"],
      "README.md": [1516, "markdown", "Oct 12 09:06"],
      "package.json": [347, "json", "Oct 9 13:10"]
    };

    const after  = {
      "app.js": [516, "javascript", "Oct 13 15:12"],
      "Dockerfile": [2090, "plaintext", "Oct 13 15:12"],
      "README.md": [1516, "markdown", "Oct 12 09:06"],
      "package.json": [347, "json", "Oct 9 13:10"]
    };

    const result = diff(before, after);

    const expectedResult = [
        ['-', 'hello_world.js'],
        ['+', 'app.js'],
        ['+', 'Dockerfile']
    ];

    expect(result).to.deep.equal(expectedResult);
  });

})

mocha.run()

File: questions/objectDiff/objectdiff.py
from unittest import TestCase, main as unitTestMain


def diff(before, after):
    # write your implementation here!
    raise NotImplementedError()


class FileSystemObjectDifferTestCase(TestCase):

    def test_diff_shows_added_and_removed_elements(self):
        before = {
            "hello_world.py": (516, "Python", "Oct 12 09:04"),
            "README.md": (1516, "Markdown", "Oct 12 09:06"),
            "requirements.txt": (347, "Plain Text", "Oct 9 13:10")
        }

        after = {
            "app.py": (516, "Python", "Oct 13 15:12"),
            "Dockerfile": (2090, "Plain Text", "Oct 13 15:12"),
            "README.md": (1516, "Markdown", "Oct 12 09:06"),
            "requirements.txt": (347, "Plain Text", "Oct 9 13:10")
        }

        result = diff(before, after)

        expected_result = [
            ['-', 'hello_world.py'],
            ['+', 'app.py'],
            ['+', 'Dockerfile']
        ]

        self.assertEqual(expected_result, result)

if __name__ == '__main__':
    unitTestMain()
